src/EventStore.Client/GossipBasedEndpointDiscoverer.cs

```diff
@@ -0,0 +1,176 @@
+				return;
+			var rnd = new Random(Guid.NewGuid().GetHashCode());
+			for (int k = i; k <= j; ++k) {
+				var index = rnd.Next(k, j + 1);
+				var tmp = arr[index];
+				arr[index] = arr[k];
+				arr[k] = tmp;
+			}
+		}
+
+		private async Task<ClusterMessages.ClusterInfo> TryGetGossipFrom(Uri gossipSeed) {
+			var url = $"{Uri.UriSchemeHttps}://{gossipSeed.Host}:{gossipSeed.Port}/gossip?format=json";
+			var response = await _client.GetStringAsync(url).ConfigureAwait(false);
+			var result = JsonSerializer.Deserialize<ClusterMessages.ClusterInfo>(response, _jsonSerializerOptions);
+			return result;
+		}
+
+		private NodeEndpoint TryDetermineBestNode(IEnumerable<ClusterMessages.MemberInfo> members,
+			NodePreference nodePreference) {
+			var notAllowedStates = new[] {
+				ClusterMessages.VNodeState.Manager,
+				ClusterMessages.VNodeState.ShuttingDown,
+				ClusterMessages.VNodeState.Shutdown,
+				ClusterMessages.VNodeState.Manager,
+				ClusterMessages.VNodeState.Shutdown,
+				ClusterMessages.VNodeState.Unknown,
+				ClusterMessages.VNodeState.ReadOnlyLeaderless,
+				ClusterMessages.VNodeState.Initializing,
+				ClusterMessages.VNodeState.CatchingUp,
+				ClusterMessages.VNodeState.ResigningLeader,
+				ClusterMessages.VNodeState.ShuttingDown,
+				ClusterMessages.VNodeState.PreLeader,
+				ClusterMessages.VNodeState.PreReplica,
+				ClusterMessages.VNodeState.PreReadOnlyReplica,
+			};
+
+			var nodes = members.Where(x => x.IsAlive)
+				.Where(x => !notAllowedStates.Contains(x.State))
+				.OrderByDescending(x => x.State)
+				.ToArray();
+
+			switch (nodePreference) {
+				case NodePreference.Random:
+					RandomShuffle(nodes, 0, nodes.Length - 1);
+					break;
+				case NodePreference.Leader:
+					nodes = nodes.OrderBy(nodeEntry => nodeEntry.State != ClusterMessages.VNodeState.Leader)
+						.ToArray();
+					RandomShuffle(nodes, 0,
+						nodes.Count(nodeEntry => nodeEntry.State == ClusterMessages.VNodeState.Follower) - 1);
+					break;
+				case NodePreference.Follower:
+					nodes = nodes.OrderBy(nodeEntry => nodeEntry.State != ClusterMessages.VNodeState.Follower)
+						.ToArray();
+					RandomShuffle(nodes, 0,
+						nodes.Count(nodeEntry => nodeEntry.State == ClusterMessages.VNodeState.Follower) - 1);
+					break;
+				case NodePreference.ReadOnlyReplica:
+					nodes = nodes.OrderBy(nodeEntry => !IsReadOnlyReplicaState(nodeEntry.State))
+						.ToArray();
+					RandomShuffle(nodes, 0,
+						nodes.Count(nodeEntry => IsReadOnlyReplicaState(nodeEntry.State)) - 1);
+					break;
+			}
+
+			var node = nodes.FirstOrDefault();
+
+			if (node == default(ClusterMessages.MemberInfo)) {
+				return null;
+			}
+
+			var nodeEndpoint = new UriBuilder(Uri.UriSchemeHttps, node.ExternalHttpIp, node.ExternalHttpPort).Uri;
+			return new NodeEndpoint(node.InstanceId, nodeEndpoint);
+		}
+
+		private bool IsReadOnlyReplicaState(ClusterMessages.VNodeState state) {
+			return state == ClusterMessages.VNodeState.ReadOnlyLeaderless
```

Comment:
`notAllowedStates` contains `ReadOnlyLeaderless` and it is filtered out in `nodes`